### PR TITLE
SIG-604: Add a new command for trace with arguments

### PR DESCRIPTION
**Pull request by gauribaraskar**
Add distributed tracing generator to sigclient tool.

### CODE REVIEW

File: pkg/trace/trace.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	spanId := uuid.NewString()
+	span["spanID"] = spanId
+	span["startTime"] = time.Now()
+	span["startTimeMillis"] = time.Now().UnixNano() / int64(time.Millisecond)
+	span["flags"] = 1
+	span["operationName"] = " HTTP GET"
+	span["duration"] = fastrand.Uint32n(1_000)
+	span["tags"] = generateTagBody(10)
+	span["process"] = generateProcessTagsBody(10)
+	span["references"] = generateReferenceBody(traceId, spanId)
```
Comment (srisub98) [inline pkg/trace/trace.go:57]: we may have to add the key `"refType"` which is either `CHILD_OF` or `FollowsFrom`

Check out this diagram: https://docs.lightstep.com/docs/understand-distributed-tracing#context 

We would need to first generate a span that has no parent spans, and represents the trace a single request. 

Then, for the first random span, we create a `refType: CHILD_OF` for the above trace id. Then all other random spans can do a `refType: FollowsFrom` the previous span id. 

This way we are synthetically generating a trace that has many sequential spans.

**Review by srisub98 — COMMENTED** [review]
Overall approach looks good, we may hit some issues during the jaeger UI testing due to missing `"refType"` keys

File: README.md [hunk 2]
```diff
@@ -48,10 +47,22 @@ Options:
 -c  continuous             If true, ignores -n and -v and will continuously send queries to the destination and will log results
 ```
 
+## Generating traces
+To generate synthetic traces: 
+```bash
+$ ./load-test traces -f test.json -t 1000 -s 10
```
Comment (srisub98) [inline README.md:53]: can you also add a note outlining how two different `./load-test ingest ...` commands are needed (one for service & one for spans)?

File: pkg/trace/trace.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+func randomHex(n int) string {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return ""
+	}
+	return hex.EncodeToString(bytes)
+}
+
+func StartTraceGeneration(file string, numTraces int, maxSpans int) {
```
Comment (srisub98) [inline pkg/trace/trace.go:28]: nit: this is the only function that we should export from this module, `WriteToServiceFile` and `WriteToFile` should not need to be

File: pkg/trace/trace.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	log.Println("Starting generation of ", numTraces, " traces...")
+	traceCounter := 0
+	log.Println("Opening file ", file, "...")
+	log.Println("Opening file trace_service.json...")
+	service_file, err := os.Create("trace_service.json")
```
Comment (srisub98) [inline pkg/trace/trace.go:33]: the service file name should be generated from the input filename. 

We can change the arg to mean file prefix, so if `--filePrefix=test_traces` then we would write `test_traces_service.json` and `test_traces_spans.json`

File: pkg/trace/trace.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		WriteToFile(traceRoot,w)
+		serviceAndOperationJson := getServiceAndOperationObject(traceRoot["operationName"], traceRoot["process"])
+		WriteToServiceFile(serviceAndOperationJson, service_writer)
+		spanCounter := 0
+		for spanCounter < int(randomNumberOfSpans) {
+			span,spanId := generateChildSpan(traceId, spanCounter,prevSpanId)
+			serviceAndOperationJson := getServiceAndOperationObject(span["operationName"], span["process"])
+			span["traceID"] = traceId
+			WriteToFile(span, w)
+			WriteToServiceFile(serviceAndOperationJson, service_writer)
```
Comment (srisub98) [inline pkg/trace/trace.go:59]: do we need to write an entry to the service file each span? or is once per trace enough?

File: pkg/trace/trace.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	processBody := make(map[string]interface{})
+	randomServiceId := fastrand.Uint32n(10)
+	processBody["serviceName"] = fmt.Sprintf("service-%d", randomServiceId)
+	listOfTags := make([]map[string]interface{}, 3)
+	tagsCounter := 0
+	for tagsCounter < 3 {
+		tags := make(map[string]interface{})
+		randomNodeId := fastrand.Uint32n(2_000)
+		randomPodId := fastrand.Uint32n(20_000)
+		randomUserId := fastrand.Uint32n(2_000_000)
```
Comment (srisub98) [inline pkg/trace/trace.go:160]: does this need to be in a loop if `tagsCounter == 3` at the end of the first iteration?

File: pkg/trace/trace.go (hunk 5 above)
Comment (gauribaraskar) [reply pkg/trace/trace.go:59]: The spans inside each trace can have different service IDs so we need to write it at both places

**Review by srisub98 — APPROVED** [review]

**Pull request merged**